File: sdk/src/main/java/com/cxense/cxensesdk/model/PerformanceEvent.java
package com.cxense.cxensesdk.model;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.cxense.cxensesdk.DependenciesProvider;
import com.cxense.cxensesdk.Preconditions;
import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Collection;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

/**
 * Push event description
 *
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2017-06-05).
 */

public final class PerformanceEvent extends Event {
    public static final String TIME = "time";
    public static final String USER_IDS = "userIds";
    public static final String PRND = "prnd";
    public static final String RND = "rnd";
    public static final String SITE_ID = "siteId";
    public static final String ORIGIN = "origin";
    public static final String TYPE = "type";
    public static final String SEGMENT_IDS = "segmentIds";
    public static final String CUSTOM_PARAMETERS = "customParameters";
    @SerializedName(TIME)
    private Long time;
    @SerializedName(USER_IDS)
    private List<UserIdentity> identities;
    @SerializedName(PRND)
    private String prnd;
    @SerializedName(RND)
    private String rnd;
    @SerializedName(SITE_ID)
    private String siteId;
    @SerializedName(ORIGIN)
    private String origin;
    @SerializedName(TYPE)
    private String type;
    @SerializedName(SEGMENT_IDS)
    private List<String> segments;
    @SerializedName(CUSTOM_PARAMETERS)
    private List<CustomParameter> customParameters;
    @SerializedName("consent")
    private List<String> consentOptions;

    private PerformanceEvent() {
        super(null);
    }

    private PerformanceEvent(@NonNull Builder builder, @NonNull List<String> consentOptions) {
        super(builder.eventId);
        rnd = String.format(Locale.US, "%d%d", Calendar.getInstance().getTimeInMillis(), (int) (Math.random() * 10E8));

        time = builder.time;
        identities = Collections.unmodifiableList(builder.identities);
        prnd = builder.prnd;
        siteId = builder.siteId;
        origin = builder.origin;
        type = builder.type;
        segments = Collections.unmodifiableList(builder.segments);
        customParameters = Collections.unmodifiableList(builder.customParameters);
        this.consentOptions = consentOptions;
    }

    /**
     * Gets the exact datetime of an event
     *
     * @return datetime of an event
     */
    @SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
    @Nullable
    public Date getTime() {
        return time != null ? new Date(TimeUnit.SECONDS.toMillis(time)) : null;
    }

    /**
     * Gets list of known user identities to identify the user.
     *
     * @return user identities
     */
    @SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
    @NonNull
    public List<UserIdentity> getIdentities() {
        return identities;
    }

    /**
     * Gets an alternative specification for page view event id.
     *
     * @return prnd
     */
    @SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
    @Nullable
    public String getPrnd() {
        return prnd;
    }

    /**
     * Gets an alternative specification for eventId.
     *
     * @return rnd
     */
    @SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
    @NonNull
    public String getRnd() {
        return rnd;
    }

    /**
     * Gets the analytics site identifier to be associated with the events.
     *
     * @return site identifier
     */
    @SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
    @NonNull
    public String getSiteId() {
        return siteId;
    }

    /**
     * Gets origin, that various DMP applications used by the customer.
     *
     * @return origin
     */
    @SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
    @NonNull
    public String getOrigin() {
        return origin;
    }

    /**
     * Gets event type, e.g., "click", "impression", "conversion", etc.
     *
     * @return event type
     */
    @SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
    @NonNull
    public String getType() {
        return type;
    }

    /**
     * Gets optional collection of matching segments to be reported.
     *
     * @return matching segments
     */
    @SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
    public List<String> getSegments() {
        return segments;
    }

    /**
     * Gets optional collection of customer-defined parameters to event.
     *
     * @return customer-defined parameters
     */
    @SuppressWarnings({"UnusedDeclaration", "WeakerAccess"}) // Public API.
    public List<CustomParameter> getCustomParameters() {
        return customParameters;
    }

    /**
     * Helper class for building {@link PerformanceEvent}
     */
    @SuppressWarnings({"UnusedDeclaration", "WeakerAccess, UnusedReturnValue"}) // Public API.
    public static class Builder {
        private String eventId;
        private Long time;
        private List<UserIdentity> identities;
        private String prnd;
        private String siteId;
        private String origin;
        private String type;
        private List<String> segments;
        private List<CustomParameter> customParameters;

        /**
         * Initialize Builder with required parameters
         *
         * @param identities List of known user identities to identify the user. Note that different users must be fed
         *                   as different events.
         * @param siteId     The analytics site identifier to be associated with the events.
         * @param origin     Differentiates various DMP applications used by the customer. Must be prefixed by the
         *                   customer prefix.
         * @param type       Differentiates various event types, e.g., "click", "impression", "conversion", etc.
         */
        public Builder(@NonNull Collection<UserIdentity> identities, @NonNull String siteId, @NonNull String origin,
                       @NonNull String type) {
            this.identities = new ArrayList<>();
            this.segments = new ArrayList<>();
            this.customParameters = new ArrayList<>();
            addIdentities(identities);
            setSiteId(siteId);
            setOrigin(origin);
            setType(type);
        }

        void setTime(long milliseconds) {
            this.time = TimeUnit.MILLISECONDS.toSeconds(milliseconds);
        }

        /**
         * Sets custom event id, that used for tracking locally.
         *
         * @param eventId event id
         * @return Builder instance
         */
        @NonNull
        public Builder setEventId(@Nullable String eventId) {
            this.eventId = eventId;
            return this;
        }

        /**
         * Sets the exact datetime of an event
         *
         * @param date datetime object
         * @return Builder instance
         */
        @NonNull
        public Builder setTime(@NonNull Date date) {
            setTime(date.getTime());
            return this;
        }

        /**
         * Sets current datetime as exact datetime of an event
         *
         * @return Builder instance
         */
        @NonNull
        public Builder setCurrentTime() {
            setTime(System.currentTimeMillis());
            return this;
        }

        /**
         * Adds known user identity to identify the user.
         *
         * @param identity {@link UserIdentity} instance
         * @return Builder instance
         * @deprecated Use {@link #addIdentities(UserIdentity...)}
         */
        @NonNull
        @Deprecated
        public Builder addIdentity(@NonNull UserIdentity identity) {
            return addIdentities(identity);
        }

        /**
         * Adds collection of known user identities to identify the user.
         *
         * @param identities {@link Collection} of {@link UserIdentity} objects
         * @return Builder instance
         */
        public Builder addIdentities(@NonNull Collection<UserIdentity> identities) {
            Preconditions.checkForNull(identities, "identities");
            for (UserIdentity identity : identities) {
                Preconditions.checkForNull(identity, "identity");
                this.identities.add(identity);
            }
            return this;
        }

        /**
         * Adds known user identities to identify the user.
         *
         * @param identities Some {@link UserIdentity} objects
         * @return Builder instance
         * @since 1.6.1
         */
        @NonNull
        public Builder addIdentities(@NonNull UserIdentity... identities) {
            return addIdentities(Arrays.asList(identities));
        }

        /**
         * Sets an alternative specification for page view event id.
         * In order to link DMP events to page views this value must be identical to the rnd value of the page view
         * event.
         *
         * @param prnd A value uniquely identifying the page request
         * @return Builder instance
         */
        @NonNull
        public Builder setPrnd(@Nullable String prnd) {
            this.prnd = prnd;
            return this;
        }

        /**
         * Sets the analytics site identifier to be associated with the events.
         *
         * @param siteId site identifier
         * @return Builder instance
         */
        @NonNull
        public Builder setSiteId(@NonNull String siteId) {
            Preconditions.checkStringForNullOrEmpty(siteId, "siteId");
            this.siteId = siteId;
            return this;
        }

        /**
         * Differentiates various DMP applications used by the customer. Must be prefixed by the customer prefix.
         *
         * @param origin prefixed by the customer prefix value
         * @return Builder instance
         */
        @NonNull
        public Builder setOrigin(@NonNull String origin) {
            Preconditions.checkStringForRegex(origin, "origin", "\\w{3}-[\\w-]+", "'%s' must be " +
                    "prefixed by the customer prefix.", "origin");
            this.origin = origin;
            return this;
        }

        /**
         * Differentiates various event types, e.g., "click", "impression", "conversion", etc.
         *
         * @param type event type
         * @return Builder instance
         */
        @NonNull
        public Builder setType(@NonNull String type) {
            Preconditions.checkStringForNullOrEmpty(type, "type");
            this.type = type;
            return this;
        }

        /**
         * Adds optional collection of matching segments to be reported.
         *
         * @param segments matching segments
         * @return Builder instance
         */
        @NonNull
        public Builder addSegments(@NonNull Collection<String> segments) {
            this.segments.addAll(segments);
            return this;
        }

        /**
         * Adds optional collection of matching segments to be reported.
         *
         * @param segments matching segments
         * @return Builder instance
         * @since 1.6.1
         */
        @NonNull
        public Builder addSegments(@NonNull String... segments) {
            return addSegments(Arrays.asList(segments));
        }

        /**
         * Adds customer-defined parameter to event.
         *
         * @param parameter {@link CustomParameter} object
         * @return Builder instance
         * @deprecated Use {@link #addCustomParameters(CustomParameter...)}
         */
        @NonNull
        @Deprecated
        public Builder addCustomParameter(@NonNull CustomParameter parameter) {
            return addCustomParameters(parameter);
        }

        /**
         * Adds optional collection of customer-defined parameters to event.
         *
         * @param parameters {@link Collection} of {@link CustomParameter} objects
         * @return Builder instance
         */
        @NonNull
        public Builder addCustomParameters(@NonNull Collection<CustomParameter> parameters) {
            Preconditions.checkForNull(parameters, "parameters");
            this.customParameters.addAll(parameters);
            return this;
        }

        /**
         * Adds optional collection of customer-defined parameters to event.
         *
         * @param parameters custom parameters
         * @return Builder instance
         * @since 1.6.1
         */
        @NonNull
        public Builder addCustomParameters(@NonNull CustomParameter... parameters) {
            return addCustomParameters(Arrays.asList(parameters));
        }

        /**
         * Build {@link PerformanceEvent} instance.
         *
         * @return PerformanceEvent instance
         */
        @NonNull
        public PerformanceEvent build() {
            return new PerformanceEvent(this, DependenciesProvider.getInstance().getCxenseConfiguration().getConsentOptionsValues());
        }
    }
}

File: sdk/src/main/java/com/cxense/cxensesdk/model/PerformanceEvent.kt
package com.cxense.cxensesdk.model

import com.cxense.cxensesdk.DependenciesProvider
import com.cxense.cxensesdk.UserProvider
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import java.util.Collections
import java.util.Date
import java.util.concurrent.TimeUnit

/**
 * Performance event description.
 * @property rnd alternative specification for eventId.
 * @property siteId The analytics site identifier to be associated with the events.
 * @property origin Differentiates various DMP applications used by the customer.
 * @property eventType Differentiates various event types, e.g., "click", "impression", "conversion", etc.
 * @property identities List of known user identities to identify the user.
 * @property eventId custom event id, that used for tracking locally.
 * @property prnd an alternative specification for page view event id.
 * @property time the exact datetime of an event
 * @property segments optional collection of matching segments to be reported.
 * @property customParameters optional collection of customer-defined parameters to event.
 */
@Suppress("unused", "MemberVisibilityCanBePrivate") // Public API.
@JsonClass(generateAdapter = true)
class PerformanceEvent internal constructor(
    eventId: String? = null,
    @Json(name = USER_IDS) val identities: List<UserIdentity>,
    @Json(name = SITE_ID) val siteId: String,
    @Json(name = ORIGIN) val origin: String,
    @Json(name = TYPE) val eventType: String,
    @Json(name = PRND) val prnd: String?,
    @Json(name = TIME) val time: Long,
    @Json(name = SEGMENT_IDS) val segments: List<String>?,
    @Json(name = CUSTOM_PARAMETERS) val customParameters: List<CustomParameter>,
    @Json(name = "consent") val consentOptions: List<String>,
    @Json(name = RND) val rnd: String = "${System.currentTimeMillis()}${(Math.random() * 10E8).toInt()}"
) : Event(eventId) {

    /**
     * @constructor Initialize Builder with required parameters
     * @property siteId The analytics site identifier to be associated with the events.
     * @property origin Differentiates various DMP applications used by the customer. Must be prefixed by the customer prefix.
     * @property eventType Differentiates various event types, e.g., "click", "impression", "conversion", etc.
     * @property identities List of known user identities to identify the user. Note that different users must be fed as different events.
     * @property eventId custom event id, that used for tracking locally.
     * @property prnd an alternative specification for page view event id. In order to link DMP events to page views this value
     * must be identical to the rnd value of the page view event.
     * @property time the exact datetime of an event in milliseconds
     * @property segments optional collection of matching segments to be reported.
     * @property customParameters optional collection of customer-defined parameters to event.
     */
    data class Builder internal constructor(
        val userProvider: UserProvider,
        var siteId: String,
        var origin: String,
        var eventType: String,
        var identities: MutableList<UserIdentity> = mutableListOf(),
        var eventId: String? = null,
        var prnd: String? = null,
        var time: Long = System.currentTimeMillis(),
        var segments: MutableList<String> = mutableListOf(),
        var customParameters: MutableList<CustomParameter> = mutableListOf()
    ) {

        @JvmOverloads
        constructor(
            siteId: String,
            origin: String,
            eventType: String,
            identities: MutableList<UserIdentity> = mutableListOf(),
            eventId: String? = null,
            prnd: String? = null,
            time: Long = System.currentTimeMillis(),
            segments: MutableList<String> = mutableListOf(),
            customParameters: MutableList<CustomParameter> = mutableListOf()
        ) : this(
            DependenciesProvider.getInstance().userProvider,
            siteId,
            origin,
            eventType,
            identities,
            eventId,
            prnd,
            time,
            segments,
            customParameters
        )

        /**
         * Adds known user identities to identify the user.
         * @param identities one or many [UserIdentity] objects
         */
        fun addIdentities(vararg identities: UserIdentity) = apply { this.identities.addAll(identities) }

        /**
         * Adds known user identities to identify the user.
         * @param identities [Iterable] with [UserIdentity] objects
         */
        fun addIdentities(identities: Iterable<UserIdentity>) = apply { this.identities.addAll(identities) }

        /**
         * Sets site identifier
         * @param siteId The analytics site identifier to be associated with the events.
         */
        fun siteId(siteId: String) = apply { this.siteId = siteId }

        /**
         * Sets event origin
         * @param origin Differentiates various DMP applications used by the customer. Must be prefixed by the customer prefix.
         */
        fun origin(origin: String) = apply { this.origin = origin }

        /**
         * Sets event type
         * @param type Differentiates various event types, e.g., "click", "impression", "conversion", etc.
         */
        fun eventType(type: String) = apply { this.eventType = type }

        /**
         * Sets custom event id
         * @param eventId custom event id, that used for tracking locally.
         */
        fun eventId(eventId: String?) = apply { this.eventId = eventId }

        /**
         * Sets page view event id
         * @param prnd an alternative specification for page view event id. In order to link DMP events to page views this value
         */
        fun prnd(prnd: String?) = apply { this.prnd = prnd }

        /**
         * Sets current datetime as datetime of an event
         */
        fun currentTime() = apply { this.time = System.currentTimeMillis() }

        /**
         * Sets datetime of an event
         * @param date the exact datetime of an event
         */
        fun time(date: Date) = apply { this.time = date.time }

        /**
         * Adds matching segments to be reported.
         * @param segments one or many segment ids
         */
        fun addSegments(vararg segments: String) = apply { this.segments.addAll(segments) }

        /**
         * Adds matching segments to be reported.
         * @param segments [Iterable] with segment ids
         */
        fun addSegments(segments: Iterable<String>) = apply { this.segments.addAll(segments) }

        /**
         * Adds custom parameters.
         * @param customParameters one or many [CustomParameter] objects
         */
        fun addCustomParameters(vararg customParameters: CustomParameter) =
            apply { this.customParameters.addAll(customParameters) }

        /**
         * Adds custom parameters.
         * @param customParameters [Iterable] with [CustomParameter] objects
         */
        fun addCustomParameters(customParameters: Iterable<CustomParameter>) =
            apply { this.customParameters.addAll(customParameters) }

        /**
         * Builds performance event
         * @throws [IllegalArgumentException] if constraints failed
         */
        fun build(): PerformanceEvent {
            check(siteId.isNotEmpty()) {
                "Site id can't be empty"
            }
            check(eventType.isNotEmpty()) {
                "Event type can't be empty"
            }
            check(origin.matches(ORIGIN_REGEX.toRegex())) {
                "Origin must be prefixed by the customer prefix."
            }

            val userIds = identities.firstOrNull {
                it.type == UserIdentity.CX_USER_TYPE
            }?.let {
                identities
            } ?: identities + UserIdentity(UserIdentity.CX_USER_TYPE, userProvider.userId)

            return PerformanceEvent(
                eventId,
                Collections.unmodifiableList(userIds),
                siteId,
                origin,
                eventType,
                prnd,
                TimeUnit.MILLISECONDS.toSeconds(time),
                segments.takeUnless { it.isEmpty() }?.let { Collections.unmodifiableList(it) },
                Collections.unmodifiableList(customParameters),
                DependenciesProvider.getInstance().cxenseConfiguration.consentSettings.consents
            )
        }
    }

    companion object {
        const val ORIGIN_REGEX = "\\w{3}-[\\w-]+"
        const val TIME = "time"
        const val USER_IDS = "userIds"
        const val PRND = "prnd"
        const val RND = "rnd"
        const val SITE_ID = "siteId"
        const val ORIGIN = "origin"
        const val TYPE = "type"
        const val SEGMENT_IDS = "segmentIds"
        const val CUSTOM_PARAMETERS = "customParameters"
    }
}
